### DISCUSSION TITLE
[HOWTO] how to set --num-prompts when benchmarking

### DISCUSSION THREAD

**Discussion by lxww302** [sgl-project/sglang#816] [Q&A] [answered]
In the [benchmark tutorial](https://github.com/sgl-project/sglang/blob/main/benchmark/blog_v0_2/README.md), we can see that the hyper-parameter `num-prompt` changes according to `--random-input`, `--random-output` and `--request-rate`, and a large difference can be observed if we change `num-prompt`. For example, if i increase `num-prompt` from 300 to 3200, the Median E2E Latency goes up from 22682.12ms to 105282.00ms. Is there any intuitions on how to chose this value?

**Comment by hnyls2002** [comment]
@lxww302 Thanks for bringing this. The parameter `--num-prompts` is decided by many ingredients. 

The capacity of KV cache is limited, so the input length and the output length actually decide how many requests can be running at the same time, based on this, you can choose the number of prompts like this:

- If you want to test an offline case, the number of prompts should be much larger than your machine's prompts capacity mentioned before.
- If you want to test an online case, you have to promise the normal serving status occupies the most of the time.

**Reply by lxww302** [reply]
@hnyls2002 Does that mean in the offline case, the larger `num_prompts` is, the higher the throughput is? But in my evaluation, I served mistralai/Mistral-7B-Instruct-v0.3 on 1 H100, and use this command to get output token throughput `python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts $NUM_PROMPTS --random-input 1024 --random-output 1024 --output-file offline.jsonl`. When changing the NUM_PROMPTS from 2000 to 4000 to 6000, the output token throughput decrease from 6712.68 to 6016.10 to 5269.99. I wonder what goes wrong with my evaluation.

**Reply by hnyls2002** [reply]
@lxww302 No, that does not mean that the larger the num prompt is, the higher the throughput is. Your result is expected. As for why the throughput decreases, it may be related to several issues, including the distribution of the random data, warmup/cool-down variance and scheduling overhead.

**Reply by hnyls2002** [reply]
@lxww302 You can try with `--disable-radix-cache` when launching the server.

**Reply by lxww302** [reply]
@hnyls2002 I still cannot understand why the larger number of prompt is, the lower the throughput is. Large number of prompts can batchify better until it reaches the performance boundary. so should the throughput increase w.r.t num_prompts until it flattens ?

**Reply by hnyls2002** [reply]
@lxww302 Have you disabled the radix cache ?

**Reply by lxww302** [reply]
@hnyls2002 yes, the command I used for starting the server is `python3 -m sglang.launch_server --model-path mistralai/Mistral-7B-Instruct-v0.3 --disable-radix-cache`, the SGlang version is 0.2.7

**Comment by Ying1123** [comment, marked as answer]
If the request rate exceeds server's capacity, the request will be queue up. Then larger number of prompts cause higher E2E latency.
